packages/ui5-tooling-modules/lib/utils/WebComponentRegistry.js

```diff
@@ -341,39 +350,67 @@ class RegistryEntry {
 				};
 			}
 		};
-		const typeDef = {
-			origType: typeInfo?.text,
-			types: [],
-		};
 
-		const types = typeInfo?.text?.match(/(?:Array<[^>]*>|[^|])+/g) || [];
+		// An original type can look like this: "Array<Foo|Bar>|string|undefined"
+		// we determine if we have a toplevel union type
+		// result will look like this: types = ["Array<Foo|Bar>", "string", "undefined"]
+		const types = typeInfo.text.match(/(?:Array<[^>]*>|[^|])+/g) || [];
 
 		for (let name of types) {
 			name = name.trim();
 			const arrayTypeMatch = name.match(/Array<(.*)>/i);
 			const multiple = !!arrayTypeMatch;
+			const origType = arrayTypeMatch?.[1] || name;
 
-			if (multiple) {
+			if (origType !== "undefined") {
+				// multiple === true ==> is array type (Array<Foo>), but not necessarily a union type in the array (Array<Foo|Bar>)
+				// multiple === false ==> can be a union type, e.g. "Foo | Bar | ..."!
 				typeDef.types.push({
-					origType: arrayTypeMatch[1],
+					origType,
 					multiple,
-					types: arrayTypeMatch[1]
-						.split("|")
-						.map((s) => deriveType(s.trim()))
-						.filter((t) => !!t),
+					dedicatedTypes: origType.split("|").map((s) => deriveType(s.trim())),
 				});
-			} else {
-				typeDef.types.push(deriveType(name));
 			}
+		}
+
+		// global multiple flag sits on typeDef
+		const multiple = typeDef.types[0].multiple || false;
 
-			// unpack UI5 type
-			// as UI5 can onbly handle one type (unlike TS)
-			typeDef.ui5TypeInfo ??= Object.assign({ multiple }, typeDef.types[0]);
-			if (typeDef.ui5TypeInfo.types?.length >= 1) {
-				Object.assign(typeDef.ui5TypeInfo, typeDef.ui5TypeInfo.types[0]);
+		// Special handling for Associations types
+		if (typeDef.origType === WEBC_ASSOCIATION_TYPE) {
+			typeDef.ui5TypeInfo = {
+				ui5Type: typeDef.types[0].ui5Type,
+				moduleType: typeDef.types[0].moduleType,
+				isComplexType: true,
+			};
+			return typeDef;
+		}
+
+		const hasNoUnionType = (obj) => obj.dedicatedTypes.length <= 1;
+		// Either single or multiple union type (e.g Array<type1 | type2> or type1 | type2)
+		if (!typeDef.types.every(hasNoUnionType)) {
```

Comment:
As discussed yesterday, `hasNoUnionType` and a NOT check afterwards makes it a bit confusing. I personally would not do a double negation.